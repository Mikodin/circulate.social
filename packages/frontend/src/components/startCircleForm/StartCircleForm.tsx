import { useState, useContext, Fragment, useReducer } from 'react';
import { useRouter } from 'next/router';
import UserContext from '../../state-management/UserContext';
import axios from 'axios';
import css from './startCircleForm.module.scss';

const CREATE_CIRCLE_ENDPOINT =
  'https://z3edrz53yg.execute-api.us-east-1.amazonaws.com/dev/circles/create';

const StartCircleForm = (props) => {
  const router = useRouter();
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [isPublic, setIsPublic] = useState(false);
  const { jwtToken } = useContext(UserContext);

  async function handleSubmit(event) {
    event.preventDefault();

    try {
      const createResponse = await axios.post(
        CREATE_CIRCLE_ENDPOINT,
        {
          name,
          description,
        },
        { headers: { Authorization: jwtToken } }
      );

      const circle = createResponse.data.circle;
      router.push(`/circles/${circle.id}`);
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  return (
    <Fragment>
      <form className={css.container} onSubmit={(e) => handleSubmit(e)}>
        <label>
          <b>Name</b>
        </label>
        <input
          type="text"
          placeholder="#notasexparty"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />

        <label>
          <b>Description</b>
        </label>
        <textarea
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />

        <label>
          <b>Allow this Circle to be discoverable?</b>
        </label>
        <input
          type="checkbox"
          checked={isPublic}
          onChange={(e) => setIsPublic(e.target.checked)}
        />

        <button type="submit">Create Circle</button>
      </form>
    </Fragment>
  );
};

export default StartCircleForm;
