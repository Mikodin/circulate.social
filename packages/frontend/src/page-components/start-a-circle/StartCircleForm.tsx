import { useState, Fragment } from 'react';
import { useRouter } from 'next/router';
import axios from 'axios';
import { Form, Input, Button, Alert, Radio, Select, Popover, Tabs } from 'antd';
import { InfoCircleOutlined } from '@ant-design/icons';

import css from './StartCircleForm.module.scss';

export const CREATE_CIRCLE_ENDPOINT =
  'https://z3edrz53yg.execute-api.us-east-1.amazonaws.com/dev/circles/create';

export interface Props {
  jwtToken: string;
}

const StartCircleForm = (props: Props): JSX.Element => {
  const router = useRouter();
  const [form] = Form.useForm();

  const [isCreateCircleInFlight, setIsCreateCircleInFlight] = useState(false);
  const [isCreateCircleError, setIsCreateCircleError] = useState(false);
  const [showDescriptionPopover, setShowDescriptionPopover] = useState(false);
  const [showFrequencyPopover, setShowFrequencyPopover] = useState(false);
  const [showPrivacyPopover, setShowPrivacyPopover] = useState(false);

  const { jwtToken } = props;

  interface FormValues {
    name: string;
    description: string;
    frequency: 'daily' | 'weekly' | 'biweekly' | 'monthly';
    privacy: 'private' | 'public';
  }

  const onFinish = async (formValues: FormValues): Promise<void> => {
    try {
      setIsCreateCircleInFlight(true);
      const createResponse = await axios.post(
        CREATE_CIRCLE_ENDPOINT,
        {
          ...formValues,
        },
        { headers: { Authorization: jwtToken } }
      );
      setIsCreateCircleInFlight(false);
      setIsCreateCircleError(false);
      const { circle } = createResponse.data;
      router.push(`/circles/[circleId]`, `/circles/${circle.id}`);
    } catch (error) {
      console.error(error);
      setIsCreateCircleInFlight(false);
      setIsCreateCircleError(true);
    }
  };

  // TODO have it popup on input focus
  const renderDescriptionPopover = (): JSX.Element => (
    <Popover
      title="Define your Circle"
      content={
        <div>
          <span>
            Descriptions are important. They help shape the community.
          </span>
          <br />
          <span>
            So are boundaries - they help to create a safe container for all.
          </span>
          <br />
          <span>
            We will surface this description to newcomers as they are joining
            the Circle. So define what you want and don't want.
          </span>
          <br />
          <span>
            What sorts of content would you <strong>like</strong> to see?
          </span>
          <br />
          <span>
            What sorts of content would you <strong>not</strong> like to see?
          </span>
          <br />
          <span>Who is this for? Who is this not for?</span>
          <br />
          <span>Get clear - be specific - check out the examples below!</span>
          <br />
          <Tabs defaultActiveKey="1" style={{ maxWidth: 'fit-content' }}>
            <TabPane tab="Example 1" key="1">
              Some good example of a description
            </TabPane>
            <TabPane tab="Example 2" key="2">
              Another good example of a description
            </TabPane>
            <TabPane tab="Example 3" key="3">
              Yet another good example of a description
            </TabPane>
          </Tabs>
        </div>
      }
      trigger="click"
      onVisibleChange={(): void =>
        setShowDescriptionPopover(!showDescriptionPopover)
      }
      visible={showDescriptionPopover}
    >
      <InfoCircleOutlined
        onClick={(): void => setShowDescriptionPopover(!showDescriptionPopover)}
      />
    </Popover>
  );
  const renderFrequencyPopover = (): JSX.Element => (
    <Popover
      title="Newsletters only go out when there is content"
      content={
        <div>
          <span>Daily: Once a day</span>
          <br />
          <span>Weekly: Once a week</span>
          <br />
          <span>Bi-Weekly: Every other week</span>
          <br />
          <span>Monthly: Once per month</span>
        </div>
      }
      trigger="click"
      onVisibleChange={(): void =>
        setShowFrequencyPopover(!showFrequencyPopover)
      }
      visible={showFrequencyPopover}
    >
      <InfoCircleOutlined
        onClick={(): void => setShowFrequencyPopover(!showFrequencyPopover)}
      />
    </Popover>
  );

  const renderPrivacyPopover = (): JSX.Element => (
    <Popover
      title="On Circle privacy"
      content={
        <div>
          <span>
            At the moment there is no public way of discovering a Circle without
            being invited.
          </span>
          <br />
          <span>
            We are planning to create a public facing Circle discovery feed.
          </span>
          <br />
          <span>
            If your Circle is &quot;public&quot; it will be surfaced in that
            discovery feed.
          </span>
        </div>
      }
      trigger="click"
      onVisibleChange={(): void => setShowPrivacyPopover(!showPrivacyPopover)}
      visible={showPrivacyPopover}
    >
      <InfoCircleOutlined
        onClick={(): void => setShowPrivacyPopover(!showPrivacyPopover)}
      />
    </Popover>
  );
  const { TabPane } = Tabs;
  return (
    <Fragment>
      <Form
        className={css.form}
        form={form}
        name="horizontal_login"
        size="large"
        layout="vertical"
        onFinish={onFinish}
        initialValues={
          {
            name: '',
            description: '',
            frequency: 'weekly',
            privacy: 'private',
          } as FormValues
        }
        onValuesChange={(): void => {
          setIsCreateCircleError(false);
        }}
      >
        <Form.Item
          name="name"
          rules={[
            { required: true, message: "Please input your Circle's name" },
          ]}
          label="Name of Circle"
        >
          <Input placeholder="Circle name" type="text" />
        </Form.Item>

        <Form.Item
          name="description"
          rules={[
            {
              required: false,
              message: 'Please input a description for your Circle',
            },
          ]}
          label={<span>{renderDescriptionPopover()} Circle description</span>}
        >
          <Input.TextArea
            rows={3}
            placeholder="Circle description"
            onFocus={(): void => setShowDescriptionPopover(true)}
            onBlur={(): void => setShowDescriptionPopover(false)}
            style={{ fontSize: '16px' }}
          />
        </Form.Item>
        <Form.Item
          name="frequency"
          label={<span>{renderFrequencyPopover()} Cirulation frequency</span>}
          hasFeedback
          rules={[
            {
              required: true,
              message:
                'Please select how often you would like us to send your newsletter',
            },
          ]}
        >
          <Select
            placeholder="Please select your Circulation frequency"
            onFocus={(): void => setShowFrequencyPopover(true)}
            onBlur={(): void => setShowFrequencyPopover(false)}
          >
            <Select.Option value="daily">Daily</Select.Option>
            <Select.Option value="weekly">Weekly</Select.Option>
            <Select.Option value="biweekly">Bi-weekly</Select.Option>
            <Select.Option value="monthly">Monthly</Select.Option>
          </Select>
        </Form.Item>
        <Form.Item
          name="privacy"
          label={<span>{renderPrivacyPopover()} Circle Privacy</span>}
        >
          <Radio.Group>
            <Radio
              className={css.verticalRadioButton}
              value="private"
              // @ts-expect-error
              onFocus={(): void => setShowPrivacyPopover(true)}
              onBlur={(): void => setShowPrivacyPopover(false)}
            >
              Private
            </Radio>
            <Radio
              className={css.verticalRadioButton}
              value="public"
              // @ts-expect-error
              onFocus={(): void => setShowPrivacyPopover(true)}
              onBlur={(): void => setShowPrivacyPopover(false)}
            >
              Public
            </Radio>
          </Radio.Group>
        </Form.Item>
        {isCreateCircleError && (
          <Alert
            message="Something went wrong, please try again."
            type="error"
            showIcon
            closable
            banner
            style={{ marginBottom: '10px' }}
          />
        )}
        <Form.Item shouldUpdate={true}>
          {(): JSX.Element => {
            const isNameTouched = form.isFieldTouched('name');
            const isFormTouched = isNameTouched;
            return (
              <Button
                data-testid="submitButton"
                loading={isCreateCircleInFlight}
                type="primary"
                htmlType="submit"
                size="large"
                block
                disabled={Boolean(
                  !isFormTouched ||
                    form.getFieldsError().filter(({ errors }) => errors.length)
                      .length
                )}
              >
                {isCreateCircleInFlight
                  ? 'Creating Circle...'
                  : 'Create Circle'}
              </Button>
            );
          }}
        </Form.Item>
      </Form>
    </Fragment>
  );
};

export default StartCircleForm;
