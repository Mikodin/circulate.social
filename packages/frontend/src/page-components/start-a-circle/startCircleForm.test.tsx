import {
  render,
  RenderResult,
  act,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import axios from 'axios';

import StartCircleForm, {
  Props,
  CREATE_CIRCLE_ENDPOINT,
} from './StartCircleForm';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const mockRouterPushSpy = jest.fn(() => true);
jest.mock('next/router', () => ({
  // eslint-disable-next-line
  useRouter() {
    return {
      push: mockRouterPushSpy,
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
    };
  },
}));

const MOCK_JWT_TOKEN = '123-abc';
const DEFAULT_PROPS = { jwtToken: MOCK_JWT_TOKEN };
function renderStartCircleForm(props?: Partial<Props>): RenderResult {
  return render(<StartCircleForm {...DEFAULT_PROPS} {...props} />);
}

// eslint-disable-next-line
function getAllInputs(container: RenderResult) {
  const { queryByPlaceholderText, queryByText } = container;
  const inputNameOfCircle = queryByPlaceholderText(/Circle name/i);
  const inputDescription = queryByPlaceholderText(/Circle description/i);
  const dropdownCircleFrequency = queryByPlaceholderText(
    /Please select your Circulation frequency/i
  );
  const radioCirclePrivate = queryByText(/Private/i);
  const radioCirclePublic = queryByText(/Public/i);
  const buttonSubmit = queryByText(/Create Circle/i);
  return {
    inputNameOfCircle,
    inputDescription,
    buttonSubmit,
    dropdownCircleFrequency,
    radioCirclePrivate,
    radioCirclePublic,
  };
}

const nameOfCircleValue = 'The greatest circle ever';
const descriptionValue = 'The greatest description ever';

const setupCompleteForm = async (): Promise<RenderResult> => {
  const container = renderStartCircleForm();

  const {
    inputNameOfCircle,
    inputDescription,
    // circleFrequencyDropdown,
    // circlePrivacyRadio,
  } = getAllInputs(container);

  await act(async () => {
    fireEvent.change(inputNameOfCircle, {
      target: { value: nameOfCircleValue },
    });

    fireEvent.change(inputDescription, {
      target: { value: descriptionValue },
    });
  });

  return container;
};

describe('StartCircleForm', () => {
  beforeEach(() => {
    mockedAxios.post.mockClear();
    mockRouterPushSpy.mockClear();
  });

  it('Should render', () => {
    expect(renderStartCircleForm()).toMatchSnapshot();
  });

  describe('On submission of a complete form', () => {
    it('Should call Axios.post to the API url with all the form fields and props.jwtToken in the headers', async () => {
      const container = await setupCompleteForm();
      const { buttonSubmit } = getAllInputs(container);
      const data = { data: { circle: { id: '123' } } };
      mockedAxios.post.mockImplementationOnce(() => Promise.resolve(data));
      await act(async () => {
        fireEvent.submit(buttonSubmit);
      });

      await waitFor(() => {
        expect(mockedAxios.post).toBeCalledWith(
          CREATE_CIRCLE_ENDPOINT,
          {
            name: nameOfCircleValue,
            description: descriptionValue,
            frequency: 'weekly',
            privacy: 'private',
          },
          { headers: { Authorization: MOCK_JWT_TOKEN } }
        );
      });
    });

    it('Should route the user to the newly created circles page', async () => {
      const container = await setupCompleteForm();
      const { buttonSubmit } = getAllInputs(container);
      const data = { data: { circle: { id: '123' } } };
      mockedAxios.post.mockImplementationOnce(() => Promise.resolve(data));
      await act(async () => {
        fireEvent.submit(buttonSubmit);
      });

      await waitFor(() => {
        expect(mockRouterPushSpy).toBeCalledWith(
          '/circles/[circleId]',
          `/circles/${data.data.circle.id}`
        );
      });
    });

    it('should display "Creating Circle..." while the request is in flight', async () => {
      const container = await setupCompleteForm();
      const { queryByText } = container;
      const { buttonSubmit } = getAllInputs(container);
      const data = { data: { circle: { id: '123' } } };
      mockedAxios.post.mockImplementationOnce(() => Promise.resolve(data));

      act(() => {
        fireEvent.submit(buttonSubmit);
      });

      await waitFor(() => {
        expect(queryByText(/Creating Circle.../i)).toBeTruthy();
      });
    });
  });

  describe('Unhappy path', () => {
    describe(`When the "name" field isn't filled out`, () => {
      it('should disable the submit button', async () => {
        const container = await renderStartCircleForm();
        const { buttonSubmit } = getAllInputs(container);
        act(() => {
          fireEvent.submit(buttonSubmit);
        });

        await waitFor(() => {
          expect(mockedAxios).not.toBeCalled();
        });
      });
    });
    describe(`When the POST request returns an error`, () => {
      it('should display an alert to the user telling them to try again', async () => {
        const container = await setupCompleteForm();
        const { queryByText } = container;
        const { buttonSubmit } = getAllInputs(container);

        mockedAxios.post.mockImplementationOnce(() => Promise.reject());

        await act(async () => {
          fireEvent.submit(buttonSubmit);
        });

        await waitFor(() => {
          expect(
            queryByText(/Something went wrong, please try again./i)
          ).toBeTruthy();
        });
      });
    });
  });
});
