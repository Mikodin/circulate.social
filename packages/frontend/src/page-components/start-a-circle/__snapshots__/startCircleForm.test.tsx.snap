// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartCircleForm Should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-vertical ant-form-hide-required-mark ant-form-large form"
        id="horizontal_login"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="horizontal_login_name"
              title="Name of Circle"
            >
              Name of Circle
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input ant-input-lg"
                  id="horizontal_login_name"
                  placeholder="Circle name"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="horizontal_login_description"
              title=""
            >
              <span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
                 Circle description
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <textarea
                  class="ant-input"
                  id="horizontal_login_description"
                  placeholder="Circle description"
                  rows="3"
                  style="font-size: 16px;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="horizontal_login_frequency"
              title=""
            >
              <span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
                 Cirulation frequency
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-lg ant-select-single ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="horizontal_login_frequency_list_0"
                        aria-autocomplete="list"
                        aria-controls="horizontal_login_frequency_list"
                        aria-haspopup="listbox"
                        aria-owns="horizontal_login_frequency_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="horizontal_login_frequency"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-item"
                    >
                      Weekly
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="horizontal_login_privacy"
              title=""
            >
              <span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
                 Circle Privacy
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-radio-group ant-radio-group-outline ant-radio-group-large"
                  id="horizontal_login_privacy"
                >
                  <label
                    class="verticalRadioButton ant-radio-wrapper ant-radio-wrapper-checked"
                  >
                    <span
                      class="ant-radio ant-radio-checked"
                    >
                      <input
                        checked=""
                        class="ant-radio-input"
                        type="radio"
                        value="private"
                      />
                      <span
                        class="ant-radio-inner"
                      />
                    </span>
                    <span>
                      Private
                    </span>
                  </label>
                  <label
                    class="verticalRadioButton ant-radio-wrapper"
                  >
                    <span
                      class="ant-radio"
                    >
                      <input
                        class="ant-radio-input"
                        type="radio"
                        value="public"
                      />
                      <span
                        class="ant-radio-inner"
                      />
                    </span>
                    <span>
                      Public
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-lg ant-btn-block"
                  data-testid="submitButton"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Create Circle
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-vertical ant-form-hide-required-mark ant-form-large form"
      id="horizontal_login"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="horizontal_login_name"
            title="Name of Circle"
          >
            Name of Circle
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-lg"
                id="horizontal_login_name"
                placeholder="Circle name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="horizontal_login_description"
            title=""
          >
            <span>
              <span
                aria-label="info-circle"
                class="anticon anticon-info-circle"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="info-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
               Circle description
            </span>
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <textarea
                class="ant-input"
                id="horizontal_login_description"
                placeholder="Circle description"
                rows="3"
                style="font-size: 16px;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="horizontal_login_frequency"
            title=""
          >
            <span>
              <span
                aria-label="info-circle"
                class="anticon anticon-info-circle"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="info-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
               Cirulation frequency
            </span>
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-lg ant-select-single ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="horizontal_login_frequency_list_0"
                      aria-autocomplete="list"
                      aria-controls="horizontal_login_frequency_list"
                      aria-haspopup="listbox"
                      aria-owns="horizontal_login_frequency_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="horizontal_login_frequency"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-item"
                  >
                    Weekly
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="horizontal_login_privacy"
            title=""
          >
            <span>
              <span
                aria-label="info-circle"
                class="anticon anticon-info-circle"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="info-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
               Circle Privacy
            </span>
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-radio-group ant-radio-group-outline ant-radio-group-large"
                id="horizontal_login_privacy"
              >
                <label
                  class="verticalRadioButton ant-radio-wrapper ant-radio-wrapper-checked"
                >
                  <span
                    class="ant-radio ant-radio-checked"
                  >
                    <input
                      checked=""
                      class="ant-radio-input"
                      type="radio"
                      value="private"
                    />
                    <span
                      class="ant-radio-inner"
                    />
                  </span>
                  <span>
                    Private
                  </span>
                </label>
                <label
                  class="verticalRadioButton ant-radio-wrapper"
                >
                  <span
                    class="ant-radio"
                  >
                    <input
                      class="ant-radio-input"
                      type="radio"
                      value="public"
                    />
                    <span
                      class="ant-radio-inner"
                    />
                  </span>
                  <span>
                    Public
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-lg ant-btn-block"
                data-testid="submitButton"
                disabled=""
                type="submit"
              >
                <span>
                  Create Circle
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
